android/tv-device-auth-grant/app/src/main/java/com/ttd/sso/sample/MainActivity.kt

```diff
@@ -0,0 +1,167 @@
+            text = stringResource(R.string.welcome_description),
+            style = MaterialTheme.typography.bodyLarge,
+        )
+
+        Spacer(Modifier.height(32.dp))
+
+        when (viewState) {
+            is SignedInState -> {
+                Button(
+                    onClick = viewModel::signOut,
+                    modifier = Modifier
+                        .focusRequester(focusRequester)
+                        .sizeIn(minWidth = 240.dp, minHeight = 40.dp),
+                ) {
+                    Text(text = stringResource(R.string.action_sign_out))
+                }
+            }
+
+            else -> {
+                Button(
+                    onClick = viewModel::signIn,
+                    modifier = Modifier
+                        .focusRequester(focusRequester)
+                        .sizeIn(minWidth = 240.dp, minHeight = 40.dp),
+                ) {
+                    Text(text = stringResource(R.string.action_sign_in))
+                }
+            }
+        }
+
+        Spacer(Modifier.height(16.dp))
+
+        when (val state = viewState) {
+            is SignedOutState -> Text(text = stringResource(R.string.state_signed_out))
+            is SignedInState -> Text(text = stringResource(R.string.state_signed_in, state.email))
+            is ErrorState -> Text(text = stringResource(R.string.state_error, state.error))
+            else -> Unit
+        }
+
+        Spacer(Modifier.height(16.dp))
+
+        when (val state = viewState) {
+            is UserCodeAvailableState -> {
+                Column(
+                    modifier = Modifier.fillMaxWidth(),
+                    horizontalAlignment = Alignment.CenterHorizontally,
+                ) {
+                    Text(text = stringResource(R.string.code, state.userCode))
+                    Text(text = stringResource(R.string.url, state.verificationUriComplete))
+
+                    Spacer(Modifier.height(16.dp))
+
+                    Image(
+                        modifier = Modifier.size(180.dp),
+                        bitmap = getQrCode(state.verificationUriComplete),
+                        contentDescription = null,
+                    )
+                }
+            }
+
+            is UserCodeExpiredState -> {
+                Text(text = stringResource(R.string.code_expired))
+            }
+
+            else -> Unit
+        }
+    }
+
+    LaunchedEffect(Unit) {
+        focusRequester.requestFocus()
+    }
+}
+
+private const val QrCodeSize = 512
+private fun getQrCode(uri: String): ImageBitmap {
+    val bits = QRCodeWriter().encode(uri, QR_CODE, QrCodeSize, QrCodeSize)
+    return Bitmap.createBitmap(QrCodeSize, QrCodeSize, Bitmap.Config.RGB_565).also {
+        for (x in 0 until QrCodeSize) {
+            for (y in 0 until QrCodeSize) {
+                it.setPixel(x, y, if (bits[x, y]) android.graphics.Color.BLACK else android.graphics.Color.WHITE)
+            }
+        }
+    }.asImageBitmap()
+}
```

Comment:
I guess you could do this directly on a canvas instead of the Bitmap intermediate step. Up to you (sounds fun to do but to each their own haha)

Reply:
Nice idea, but I think i'd prefer to just leave as is :)